### PR TITLE
[bugfix] Fix performance component timing extraction

### LINKED ISSUE #1377: [Bug] Downstream perf CI stage metric key mapping bug
### Describe the bug

https://github.com/hao-ai-lab/FastVideo/pull/1337 introduced end to end latency and throughput regression along with introducing perf CI stage metric key mapping bug

### Reproduction

/test performance on any pr

### Environment

any version.

### PR DESCRIPTION
**Pull request by Satyam-53**
## Purpose

  Fixes #1377.

  Restore downstream performance CI component timings after pipeline stage logging moved to pipeline-
  unique stage keys.

  ## Changes

  - Map performance stage timings using each stage entry’s `stage_class`, with fallback to legacy class-
  name stage keys.
  - Add focused tests for real `PipelineLoggingInfo`, pipeline-keyed logs, legacy logs, duplicate
  denoising accumulation, and unmapped stages.
  - Update performance benchmark docs to describe pipeline-unique stage logging and component metric
  extraction.

  ## Test Plan

  ```bash
  python -m pytest \
    fastvideo/tests/performance/test_inference_performance_component_times.py \
    fastvideo/tests/performance/test_compare_baseline_policy.py \
    fastvideo/tests/performance/test_dashboard_service.py \
    fastvideo/tests/performance/test_dashboard_api.py \
    -q
```

 ```bash
  pre-commit run --files \
    docs/contributing/performance_benchmarks.md \
    fastvideo/tests/performance/test_inference_performance.py \
    fastvideo/tests/performance/test_inference_performance_component_times.py
```

  ## Test Results
  <details>
  <summary>Test output</summary>

  ........................                                                 [100%]
  24 passed in 0.96s

  yapf.................................................(no files to check)Skipped
  ruff (legacy alias)..................................(no files to check)Skipped
  codespell................................................................Passed
  PyMarkdown...............................................................Passed
  Lint GitHub Actions workflow files...................(no files to check)Skipped
  mypy.................................................(no files to check)Skipped
  Check for spaces in all filenames........................................Passed
  Suggestion...............................................................Passed
  </details>

  ## Checklist

  - [x] I ran pre-commit on changed files and fixed all issues
  - [x] I added or updated tests for my changes
  - [x] I updated documentation if needed
  - [x] I considered GPU memory impact of my changes

  For model/pipeline changes, also check:

  - [ ] I verified SSIM regression tests pass
  - [ ] I updated the support matrix if adding a new model

### CODE REVIEW

**Comment by SolitaryThinker** [comment]
**Hi @Satyam-53** — automated review from Gob, one of @SolitaryThinker's AI reviewers. Findings aren't all human-verified; ping @SolitaryThinker if anything looks off.

## TL;DR

This looks ship-ready: the extractor now reads `stage_class` from each pipeline-unique stage entry and falls back to the old class-name stage key shape, which addresses #1377's component-metric attribution break. The new focused tests cover the important compatibility paths, including real `PipelineLoggingInfo`, new keyed logs, legacy logs, duplicate denoising accumulation, and unmapped stages.

## Verdict: **approve**

- S0 (blockers): 0
- S1 (must-fix): 0
- S2 (should-fix; surfaced if `persistent` or `important`): 0
- S3 (discussion): not shown here; see `review.md`

---

## What's good

- The fix is narrowly scoped to the performance CI extraction path: `stage_data.get("stage_class", stage_name)` preserves new pipeline-unique stage keys while keeping legacy class-name logs working.
- Duplicate component classes are handled by summing into the existing component metric instead of overwriting, so repeated denoising stages still contribute to `dit_time_s` correctly.
- Unmapped stages are skipped without affecting mapped component metrics, which keeps additional pipeline bookkeeping stages from breaking performance reports.
- The docs now describe the same `stage_class`-based extraction and legacy fallback implemented by the test helper, so future benchmark debugging has the right mental model.

One robustness nit from the bot review remains technically true: malformed `stages` entries that are not dict-like would still raise on `.get()`. I do not think that should block this PR because the producer contract uses `PipelineLoggingInfo.stages: OrderedDict[str, dict[str, Any]]`, and the legacy/new shapes covered here both use dict stage entries.

---

*— Gob (@SolitaryThinker's AI reviewer). Full review (including S3 items) is archived locally.*

File: fastvideo/tests/performance/test_inference_performance.py [hunk 1]
```diff
@@ -98,10 +98,12 @@ def _extract_component_times(result: dict) -> dict[str, float | None]:
         return component_times
     logger.info("Discovered pipeline stages: %s", list(stages.keys()))
     for stage_name, stage_data in stages.items():
-        metric_key = STAGE_METRIC_MAP.get(stage_name)
+        stage_class = stage_data.get("stage_class", stage_name)
+        metric_key = STAGE_METRIC_MAP.get(stage_class)
```
[the remark this answers is not in the document]
Comment (Satyam-53) [reply fastvideo/tests/performance/test_inference_performance.py:105]: addressed.

File: fastvideo/tests/performance/test_inference_performance.py [hunk 2]
```diff
@@ -98,10 +98,15 @@ def _extract_component_times(result: dict) -> dict[str, float | None]:
         return component_times
     logger.info("Discovered pipeline stages: %s", list(stages.keys()))
     for stage_name, stage_data in stages.items():
-        metric_key = STAGE_METRIC_MAP.get(stage_name)
+        if not isinstance(stage_data, Mapping):
+            logger.debug("Skipping malformed stage '%s' data: %r", stage_name, stage_data)
+            continue
+        stage_class = stage_data.get("stage_class", stage_name)
+        metric_key = STAGE_METRIC_MAP.get(stage_class)
```
Comment (SolitaryThinker) [inline fastvideo/tests/performance/test_inference_performance.py:105]: **NIT (maintainability):** `STAGE_METRIC_MAP.get(stage_class)` is an exact match on `__class__.__name__`. Correct for the current Wan benchmark (base `DenoisingStage`/`TextEncodingStage`/`DecodingStage`), but a future benchmark whose pipeline uses a `DenoisingStage` subclass (e.g. `CosmosDenoisingStage`, `Gen3CDenoisingStage`, `CausalDenoisingStage`) would silently map to `dit_time_s=None` and skip the DiT gate — the same attribution gap this PR fixes. You already had to add `DmdDenoisingStage` explicitly, which is the tell. Consider matching by `isinstance`/base class (or normalizing a subclass to its component base) so a new config doesn't need a map edit. Not blocking — no current benchmark hits it.

File: docs/contributing/performance_benchmarks.md [hunk 3]
```diff
@@ -105,11 +105,15 @@ Each benchmark records six metrics:
...
 `test_inference_performance.py` temporarily sets `FASTVIDEO_STAGE_LOGGING=1`
 while it runs so pipeline stage execution times are available in
-`generate_video(...).logging_info`. It maps `TextEncodingStage` to
-`text_encoder_time_s`, `DenoisingStage` and `DmdDenoisingStage` to
-`dit_time_s`, and `DecodingStage` to `vae_decode_time_s`. If a pipeline does
-not report one of those stages, that component metric is stored as `null` and
-is skipped by the static threshold and rolling baseline checks.
+`generate_video(...).logging_info`. Stage logs use pipeline-unique keys such as
+`prompt_encoding_stage` so duplicate stage classes do not collide; each entry's
+`stage_class` field is mapped to component metrics. The extractor maps
```
Comment (SolitaryThinker) [inline docs/contributing/performance_benchmarks.md]: **NIT (docs):** "each entry's `stage_class` field is mapped" slightly overstates it — only the `PipelineStage`-driven entries carry `stage_class`. The generator post-processing entries (`PostDecodeFrameProcessStage`, `VideoSaveStage`, `AudioMuxStage`) are logged via `add_stage_execution_time(...)` only and have no `stage_class`. Harmless (they fall through to the `stage_name` lookup, miss the map, and are skipped — exactly what the `ignores_unmapped_stages` test asserts), but the sentence could note that those entries are intentionally ignored.

File: fastvideo/tests/performance/test_inference_performance_component_times.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    assert _extract_component_times(result) == {
+        "text_encoder_time_s": 1.2,
+        "dit_time_s": 3.4,
+        "vae_decode_time_s": 0.8,
+    }
+
+
+def test_extract_component_times_keeps_legacy_class_name_keys():
```
Comment (SolitaryThinker) [inline fastvideo/tests/performance/test_inference_performance_component_times.py:48]: **NIT (tests):** Nice coverage. One note: this legacy-class-name-keys case (and `ignores_unmapped_stages`) passes identically on the pre-fix code, so it characterizes the backward-compat fallback rather than guarding the #1377 regression. The actual regression guard is `test_extract_component_times_uses_stage_class_for_pipeline_stage_keys` (it goes all-`None` on the old `stage_name` lookup) — worth a one-line comment marking which test pins the fix vs which document fallback behavior, so a future refactor knows which one must stay red on a revert.

**Review by SolitaryThinker — COMMENTED** [review]
Hi @Satyam-53, thanks for the PR!

Traced the fix end-to-end and it correctly restores component-timing attribution for the live benchmark. Since #1337 the stages dict is keyed by pipeline-unique `stage_key`, so mapping each entry's stored `stage_class` (with the `stage_name` fallback) is the right fix. Confirmed the only benchmark config (`wan-t2v-1.3b.json` → `Wan-AI/Wan2.1-T2V-1.3B-Diffusers` → base `WanPipeline`) registers the base `TextEncodingStage`/`DenoisingStage`/`DecodingStage`, whose `__class__.__name__` match `STAGE_METRIC_MAP` exactly, so all three metrics populate. The `isinstance(stage_data, Mapping)` guard and the duplicate-class accumulation both look right, and the new tests cover the real `PipelineLoggingInfo` object plus the keyed/legacy/duplicate/unmapped/malformed shapes.

No blockers. A few small notes inline, all NIT-level.

FYI the `build` check is red on a pre-existing broken docs link (`docs/training/examples/wan_t2v_1.3B_mixkit.md:59`), unrelated to this PR.

## Dropped (considered, not flagged)
- New test only runs in the performance lane (`pytest ./fastvideo/tests/performance`), not per-PR fastcheck — dropped: `_extract_component_times` is perf-CI tooling that only executes in that lane anyway, so co-locating its guard there is correct placement, not a coverage gap.

File: fastvideo/tests/performance/test_inference_performance_component_times.py (hunk 4 above)
Comment (Satyam-53) [reply fastvideo/tests/performance/test_inference_performance_component_times.py:48]: Addressed. Added comments to distinguish the actual #1377 regression guard from the tests that document legacy fallback and intentionally ignored generator-side timings.

File: docs/contributing/performance_benchmarks.md (hunk 3 above)
Comment (Satyam-53) [reply docs/contributing/performance_benchmarks.md]: Addressed. Updated the docs to say that `stage_class` applies to `PipelineStage` entries, while
  generator-side timings like `PostDecodeFrameProcessStage`, `VideoSaveStage`, and `AudioMuxStage` are intentionally ignored.

File: fastvideo/tests/performance/test_inference_performance.py (hunk 2 above)
Comment (Satyam-53) [reply fastvideo/tests/performance/test_inference_performance.py:105]: Left unchanged for now. This PR targets the current Wan performance benchmark, whose stages match the existing map exactly. Subclass normalization would broaden behavior beyond this bugfix, so I’d handle  that separately if/when we add a benchmark using those subclasses. - Opened an issue to track this- #1475

**Review by Satyam-53 — COMMENTED** [review]
Addressed nits.

**Review by SolitaryThinker — APPROVED** [review]
Hi @Satyam-53, thanks for the PR!

The earlier nits are all handled: NIT2 (docs) and NIT3 (test comments) are resolved, and NIT1 (subclass exact-name match) is reasonably deferred to #1475 — agreed it's out of scope for this bugfix.

Re-verified the fix still holds at 52eab31b (it's unchanged since the last review): the only benchmark config (`wan-t2v-1.3b.json` → base `WanPipeline`) registers base `TextEncodingStage`/`DenoisingStage`/`DecodingStage`, whose `__class__.__name__` match `STAGE_METRIC_MAP` exactly, so all three component metrics still populate. The docs rewrite is accurate — every `PipelineStage` entry gets a `stage_class` (stages/base.py:157), and the named generator-side stages (`PostDecodeFrameProcessStage`/`VideoSaveStage`/`AudioMuxStage`) are written into the stages dict but correctly fall through as unmapped, so "intentionally ignored" is the right framing. The added test comments cleanly separate the #1377 regression guard from the legacy-fallback and generator-side-ignore cases.

LGTM.

FYI the `build` check is still red on the same pre-existing broken docs link (`docs/training/examples/wan_t2v_1.3B_mixkit.md:59`), unrelated to this PR.

## Dropped (considered, not flagged)
- Nothing new this pass. The prior note that the component-times test only runs in the performance lane (not per-PR fastcheck) still stands as a non-issue — `_extract_component_times` is perf-CI tooling that only executes in that lane, so co-locating its guard there is correct placement.
